src/services/formatting/rules.ts

```diff
@@ -780,4 +789,79 @@ namespace ts.formatting {
     function isNonNullAssertionContext(context: FormattingContext): boolean {
         return context.contextNode.kind === SyntaxKind.NonNullExpression;
     }
+
+    function isSemicolonDeletionContext(context: FormattingContext): boolean {
+        let nextTokenKind = context.nextTokenSpan.kind;
+        let nextTokenStart = context.nextTokenSpan.pos;
+        if (isTrivia(nextTokenKind)) {
+            const nextRealToken = context.nextTokenParent === context.currentTokenParent
+                ? findNextToken(
+                    context.currentTokenParent,
+                    findAncestor(context.currentTokenParent, a => !a.parent)!,
+                    context.sourceFile)
+                : context.nextTokenParent.getFirstToken(context.sourceFile);
+            if (!nextRealToken) {
+                return true;
+            }
+            nextTokenKind = nextRealToken.kind;
+            nextTokenStart = nextRealToken.getStart(context.sourceFile);
+        }
+
+        const startLine = context.sourceFile.getLineAndCharacterOfPosition(context.currentTokenSpan.pos).line;
+        const endLine = context.sourceFile.getLineAndCharacterOfPosition(nextTokenStart).line;
+        if (startLine === endLine) {
+            return nextTokenKind === SyntaxKind.CloseBraceToken
+                || nextTokenKind === SyntaxKind.EndOfFileToken;
+        }
+
+        if (nextTokenKind === SyntaxKind.SemicolonClassElement ||
+            nextTokenKind === SyntaxKind.SemicolonToken
+        ) {
+            return false;
+        }
+
+        if (context.contextNode.kind === SyntaxKind.InterfaceDeclaration ||
+            context.contextNode.kind === SyntaxKind.TypeAliasDeclaration
+        ) {
+            // Can’t remove semicolon after `foo`; it would parse as a method declaration:
+            //
+            // interface I {
+            //   foo;
+            //   (): void
+            // }
+            return !(isPropertySignature(context.currentTokenParent)

```

Comment:
this function returns true to mean "can delete", right? then I expected the predicate to be 

```ts
!isPropertySignature(current) || current.type || nextTokenKind !== SyntaxKind.OpenParenToken
```

Reply:
Ah-ha! I'm pretty sure I missed the surrounding parentheses here.